src/benchmarks.zig

```diff
@@ -1,349 +1,538 @@
+                }
+            };
+            const result_type: type = @TypeOf(try @call(.auto, benchFunction, arguments));
+            const runtime_type = blk: {
+                switch (result_type) {
+                    // single value
+                    Duration => {
+                        try is_multi_return.append(false);
+                        break :blk struct { result: u64 };
+                    },
+                    // multiple values
+                    else => {
+                        try is_multi_return.append(true);
+                        break :blk result_type;
+                    },
+                }
+            };
+            var runtimes: std.MultiArrayList(runtime_type) = .{};
+            defer runtimes.deinit(allocator);
+
+            //
             var min: u64 = math.maxInt(u64);
             var max: u64 = 0;
-            var runtime_sum: u128 = 0;
-
-            var i: usize = 0;
-            while (i < min_iterations or
-                (i < max_iterations and runtime_sum < max_time)) : (i += 1)
+            var sum: u64 = 0;
+
+            // NOTE: these are set to valid values on first iteration
+            const U = @typeInfo(runtime_type).Struct;
+            var sum_s: runtime_type = undefined;
+            var min_s: runtime_type = undefined;
+            var max_s: runtime_type = undefined;
+
+            //
+            var ran_out_of_time = false;
+            var runtime_timer = try sig.time.Timer.start();
+            var iter_count: u64 = 0;
+            while (iter_count < min_iterations or
+                (iter_count < max_iterations and ran_out_of_time)) : (iter_count += 1)
             {
-                const benchFunction = @field(B, def.name);
-                const ns_duration: sig.time.Duration = switch (@TypeOf(arg)) {
-                    void => benchFunction(),
-                    else => benchFunction(arg),
-                } catch |e| blk: {
-                    std.debug.print("{s} failed with error: {}\n", .{ def.name, e });
-                    num_errors += 1;
-                    break :blk .{ .ns = std.math.maxInt(u64) };
-                };
-
-                const runtime = try time_unit.unitsfromNanoseconds(ns_duration.asNanos());
-
-                runtimes[i] = runtime;
-                runtime_sum += runtime;
-                min = @min(runtimes[i], min);
-                max = @max(runtimes[i], max);
+                switch (result_type) {
+                    Duration => {
+                        const duration = try @call(.auto, benchFunction, arguments);
+                        const runtime = time_unit.convertDuration(duration);
+                        min = @min(runtime, min);
+                        max = @max(runtime, max);
+                        sum += runtime;
+                        try runtimes.append(allocator, .{ .result = runtime });
+                    },
+                    inline else => {
```

Comment:
Why `inline else`? My intuition is that either this is:
a. harmful because it's generating a branch for every possible type in the program (that seems bad).
b. superfluous because the set of types is already comptime-known, meaning it does nothing.

Reply:
8948a0cb3b14df8d6d5d65356aefa394a1c27ee8 - not sure, agreed should remove